src/relay/pass/type_infer.cc

```diff
@@ -454,7 +457,17 @@ class TypeInferencer::Resolver : public ExprMutator {
         update_missing_type_annotation_ &&
         !new_var->type_annotation.defined());
 
-    if (!need_update_type && !need_update_var && !need_update_call) return new_e;
+    bool need_update_fn = (
```

Comment:
We can punt this conversation to the future, but @MarisaKirisame and I feel like this style of resolver is going to continue to cause problems as we continue. We can chat further in person.